smooth/framework/functions/update_fitted_cost.py

```diff
@@ -1,149 +1,185 @@
-    # Function:
-    # fv_1*dependant_value^fv_2 + fv_3*dependant_value^fv_4 + ... fv_(n-1)*dependant_value^fv_n
-
-    # Get the dependant value (either if given as a component parameter of else it is the previous cost value).
-    dependant_value = get_dependant_value(component, financials, index)
-
-    # Get the fitting values.
-    fv = financials['fitting_value'][index]
-    # Get the number of fitting values [-].
-    n_fv = len(fv)
-
-    # The number of fitting values needs to be even, otherwise throw an error.
-    if n_fv % 2 != 0:
-        raise ValueError('In component {}, the number of fitting values is {}, but it needs to be even!'.format(
-            component['name'], n_fv))
-
-    # Cost value [EUR]
-    cost = 0
-    for i in range(int(n_fv/2)):
-        cost += fv[i*2] * dependant_value**fv[i*2 + 1]
-
-    # Save the costs [EUR].
-    financials['cost'] = cost
-
-
-def get_dependant_value(component, financials, index):
-    # Get an attribute of the component as the dependant value.
-    dependant_value = getattr(component, financials['dependant_value'][index])
-    if financials['dependant_value'][index] == 'capex':
-        # If the capex are chosen as the dependant value, the capex costs are meant.
-        dependant_value = dependant_value['cost']
-
-    return dependant_value
-
-
-
+import math
+
+
+def update_financials(component, financials):
+    # Calculate OPEX or CAPEX for this component.
+    # Params:
+    #  component: object of this component
+    #  financials: financial object of this component. This can be either the "capex" or the "opex" dict.
+    #
+    # This function is calculating a fix CAPEX and OPEX value for components where CAPEX and OPEX are dependant on
+    # certain values. The following list shows possible fitting methods. The fitting method is chosen by the CAPEX and
+    # OPEX key:
+    #
+    # "fix"      --> already the fix value, nothing has to be done
+    # "spec"     --> cost value needs to be multiplied with the dependant value
+    # "exp"      --> exponential cost fitting
+    # "poly"     --> polynomial cost fitting
+    # "free"     --> polynomial cost fitting with free choosable exponents
+    #
+    # If multiple keys are defined, the calculations are done sequentially in order.
+
+    # Check if this financial dictionary is empty. If so, nothing has to be calculated.
+    if not financials:
+        return
+
+    # If the keys are not given as a list, they are transformed to one so they can be iterated.
+    if type(financials['key']) is not list:
+        financials['key'] = [financials['key']]
+        financials['fitting_value'] = [financials['fitting_value']]
+        financials['dependant_value'] = [financials['dependant_value']]
+
+    # Loop through each CAPEX or OPEX key.
+    for this_index in range(len(financials['key'])):
+        # Get the dependant value (either if given as a component parameter of else it is the previous cost value).
```

Comment:
typo?: ... parameter or else...

Reply:
Did you mean by previous cost value if the cost is not dependend on anything so just in €? If so maybe we could write it more clearer?

Reply:
I did not change the comment but you are right, it does sounds incorrect. 
`previous cost value` means that if at `this_index` of the dict dependant value e.g. is stated 'capex' it uses the priviously calculated capex cost value when we have multiple fitting functions or if opex is dependant on the capex value

Reply:
Top, then we write that

Reply:
https://github.com/rl-institut/smooth/pull/15/commits/df7abcf4ba958837fc976f07a666d13ad7bd5cbb

Reply:
https://github.com/rl-institut/smooth/pull/15/commits/df7abcf4ba958837fc976f07a666d13ad7bd5cbb